File: Compiler/lexer.cpp
#include "lexer.hpp"
#include <stdexcept>

const std::unordered_set<std::string> KEYWORDS = {
    "var", "let", "func", "asm", "type", "enum",
    "mutating", "volatile", "static", "private", "public", "const", "atomic", "extern",
    "if", "else", "switch", "case", "for", "while", "return", "break", "continue",
    "align", "cast", "sizeOf", "defer", "unsafe",
    "true", "false", "nil", "import", "ui", "struct", "async", "await", "guard"
};

struct TokenRule {
    std::string type;
    std::regex pattern;
};

Lexer::Lexer(const std::string& sourceCode) : code(sourceCode) {
    indentStack.push(0);
}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> tokens;
    
    const std::vector<TokenRule> TOKEN_TYPES = {
        {"singleLineComment", std::regex(R"(//.*)")},
        {"multiLineComment", std::regex(R"(/\*[\s\S]*?\*/)")},
        {"attribute", std::regex(R"(\.[a-zA-Z_][a-zA-Z0-9_]*)")},
        {"identifier", std::regex(R"([a-zA-Z_][a-zA-Z0-9_]*)")},
        {"hexLiteral", std::regex(R"(0x[0-9a-fA-F]+)")},
        {"numberLiteral", std::regex(R"(\d+(\.\d+)?)")},
        {"stringLiteral", std::regex(R"("[^"]*")")},
        {"equalOp", std::regex(R"(==)")},
        {"notEqualOp", std::regex(R"(!=)")},
        {"lessThanOrEqualOp", std::regex(R"(<=)")},
        {"greaterThanOrEqualOp", std::regex(R"(>=)")},
        {"logicalAndOp", std::regex(R"(&&)")},
        {"logicalOrOp", std::regex(R"(\|\|)")},
        {"assignOp", std::regex(R"(=)")},
        {"addOp", std::regex(R"(\+)")},
        {"subOp", std::regex(R"(-)")},
        {"mulOp", std::regex(R"(\*)")},
        {"divOp", std::regex(R"(/)")},
        {"bitwiseAndOp", std::regex(R"(&)")},
        {"bitwiseOrOp", std::regex(R"(\|)")},
        {"bitwiseXorOp", std::regex(R"(\^))")},
        {"semicolon", std::regex(R"(;)")},
        {"colon", std::regex(R"(:)")},
        {"comma", std::regex(R"(,)")},
        {"leftParen", std::regex(R"(\()")},
        {"rightParen", std::regex(R"(\))")}
    };

    size_t length = code.length();

    while (pos < length) {
        if (code[pos] == '\n' || code[pos] == '\r') {
            bool isCrLf = (code[pos] == '\r' && pos + 1 < length && code[pos + 1] == '\n');
            pos += isCrLf ? 2 : 1;
            line++;
            col = 1;
            
            tokens.push_back({"newline", "\n", {line, col}});

            int indentLength = 0;
            while (pos < length && (code[pos] == ' ' || code[pos] == '\t')) {
                indentLength += (code[pos] == '\t') ? 4 : 1;
                pos++;
                col++;
            }

            if (pos < length && code[pos] != '\n' && code[pos] != '\r' && code[pos] != '/') {
                if (indentLength > indentStack.top()) {
                    indentStack.push(indentLength);
                    tokens.push_back({"indent", "INDENT", {line, col}});
                } else {
                    while (indentLength < indentStack.top()) {
                        indentStack.pop();
                        tokens.push_back({"dedent", "DEDENT", {line, col}});
                    }
                }
            }
            continue;
        }

        if (code[pos] == ' ' || code[pos] == '\t') {
            pos++;
            col++;
            continue;
        }

        bool matched = false;
        SourceLocation currentLoc = {line, col};

        for (const auto& rule : TOKEN_TYPES) {
            std::smatch match;
            if (std::regex_search(code.cbegin() + pos, code.cend(), match, rule.pattern, std::regex_constants::match_continuous)) {
                std::string text = match.str(0);

                if (rule.type != "singleLineComment" && rule.type != "multiLineComment") {
                    if (rule.type == "identifier" && KEYWORDS.find(text) != KEYWORDS.end()) {
                        tokens.push_back({"keyword", text, currentLoc});
                    } else {
                        tokens.push_back({rule.type, text, currentLoc});
                    }
                }

                pos += text.length();
                col += text.length();
                matched = true;
                break;
            }
        }

        if (!matched) {
            throw std::runtime_error("SyntaxError: Unexpected character '" + std::string(1, code[pos]) + 
                                     "' at line " + std::to_string(line) + ", col " + std::to_string(col));
        }
    }

    while (indentStack.size() > 1) {
        indentStack.pop();
        tokens.push_back({"dedent", "DEDENT", {line, col}});
    }

    tokens.push_back({"eof", "", {line, col}});
    return tokens;
}

File: Compiler/lexer.py
import re

# 1. Complete set of 37 Reserved Keywords for Shift
KEYWORDS = {
    # Declarations & Scope
    "var", "let", "func", "asm", "type", "enum",
    # Memory & Storage Modifiers
    "mutating", "volatile", "static", "private", "public", "const", "atomic", "extern",
    # Control Flow
    "if", "else", "switch", "case", "for", "while", "return", "break", "continue",
    # Hardware & Systems Operations
    "align", "cast", "sizeOf", "defer", "unsafe",
    # Literals & Directives
    "true", "false", "nil", "import", "ui", "struct", "async", "await", "guard"
}

# 2. Token definitions (Strict Regex Order)
TOKEN_TYPES = [
    # Comments (Skipped or preserved)
    ("singleLineComment", r"//.*"),
    ("multiLineComment", r"/\*[\s\S]*?\*/"),

    # Attributes (.name, .type, .address, .volatile)
    ("attribute", r"\.[a-zA-Z_][a-zA-Z0-9_]*"),

    # Identifiers & Keywords (Handled dynamically in tokenize logic)
    ("identifier", r"[a-zA-Z_][a-zA-Z0-9_]*"),

    # Literals (hexLiteral MUST COME BEFORE numberLiteral)
    ("hexLiteral", r"0x[0-9a-fA-F]+"),
    ("numberLiteral", r"\d+(\.\d+)?"),
    ("stringLiteral", r'"[^"]*"'),

    # Comparison & Logical Operators
    ("equalOp", r"=="),
    ("notEqualOp", r"!="),
    ("lessThanOrEqualOp", r"<="),
    ("greaterThanOrEqualOp", r">="),
    ("logicalAndOp", r"&&"),
    ("logicalOrOp", r"\|\|"),

    # Arithmetic & Bitwise Operators
    ("assignOp", r"="),
    ("addOp", r"\+"),
    ("subOp", r"-"),
    ("mulOp", r"\*"),
    ("divOp", r"/"),
    ("bitwiseAndOp", r"&"),
    ("bitwiseOrOp", r"\|"),
    ("bitwiseXorOp", r"\^"),

    # Delimiters & Symbols
    ("semicolon", r";"),
    ("colon", r":"),
    ("comma", r","),
    ("leftBrace", r"\{"),
    ("rightBrace", r"\}"),
    ("leftParen", r"\("),
    ("rightParen", r"\)"),

    # Whitespace (Skipped by Lexer)
    ("whiteSpace", r"[ \t\n\r]+"),
]

def tokenize(code):
    pos = 0
    tokens = []
    
    while pos < len(code):
        match = None
        for token_type, regex in TOKEN_TYPES:
            pattern = re.compile(regex)
            match = pattern.match(code, pos)
            
            if match:
                text = match.group(0)
                
                # Skip comments and whitespace
                if token_type not in ("whiteSpace", "singleLineComment", "multiLineComment"):
                    if token_type == "identifier" and text in KEYWORDS:
                        tokens.append(("keyword", text))
                    else:
                        tokens.append((token_type, text))
                        
                pos = match.end()
                break
                
        if not match:
            raise SyntaxError(f"Unexpected character '{code[pos]}' at position {pos}")
            
    return tokens
